### DISCUSSION TITLE
How to integrate Quarkus with JetBrains Exposed?

### DISCUSSION THREAD

**Discussion by micheljung** [quarkusio/quarkus#42593] [Design Discussions]
I'm trying to integrate Quarkus JTA with JetBrains Exposed, see also: [Integrate with Quarkus/JTA](https://youtrack.jetbrains.com/issues?preview=EXPOSED-490)

The problem is basically how to inject a Exposed-aware transaction manager into Narayana Transaction Manager, like it is somehow done for Hibernate ORM:

![image](https://github.com/user-attachments/assets/7fabaf3e-0ab1-4b35-be2d-4ce75c26e94f)
https://quarkus.io/guides/hibernate-orm#automatic-integration

I tried to copy from https://github.com/quarkusio/quarkus/tree/9070ad725ed552de597a1683c6f864f856ede3b9/extensions/hibernate-orm but I was unable to find the code that does it.

Any hints?

**Comment by micheljung** [comment]
Found this: https://github.com/quarkusio/quarkus/blob/9070ad725ed552de597a1683c6f864f856ede3b9/extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/customized/QuarkusJtaPlatform.java#L41

I don't understand this yet but it seems to be the other way around; the Hibernate ORM extension delegates to JTA?

**Comment by mschorsch** [comment]
@micheljung It should work out-of-the-box If you inject an AgroalDataSource (https://quarkus.io/guides/datasource#consuming-the-datasource) and use the DataSource as input for Exposed because the DataSource ist Transaction aware. Something like this: `Database.connect(dataSource)`. We do something similar with JOOQ and it works great.

**Comment by mschorsch** [comment]
The only thing that could be problematic ist that Exposed ist using/creating it's own Transaction Manager...

**Reply by micheljung** [reply]
Hi @mschorsch 

Exactly. The issue is that Expose checks whether there is a running transaction or not:
```
@Transactional
fun doSomething() {
  MyTable.selectAll()
}
```

throws:

    java.lang.IllegalStateException - No transaction in context.

Here https://github.com/JetBrains/Exposed/blob/8726aa890a4ffddb0f59fde9d2eb07d0a71841b4/exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/transactions/TransactionApi.kt#L229

**Reply by florentsorel** [reply]
@micheljung Did you resolve your issue? Because I have the same error. I don't know how to setup the transaction context. 

```kotlin
@ApplicationScoped
class DatabaseConfig {
    private val logger = Logger.getLogger(DatabaseConfig::class.java)

    @Inject
    @ConfigProperty(name = "quarkus.datasource.jdbc.url")
    private lateinit var jdbcUrl: String

    @Inject
    @ConfigProperty(name = "quarkus.datasource.username")
    private lateinit var username: String

    @Inject
    @ConfigProperty(name = "quarkus.datasource.password")
    private lateinit var password: String

    fun onStart(@Observes event: StartupEvent) {
        try {
            Database.connect(jdbcUrl, driver = "org.postgresql.Driver", username, password)
            logger.info("Database connection established successfully")
        } catch (e: Exception) {
            logger.error("Failed to establish database connection", e)
        }
    }
}

class ActorRepositoryImpl : ActorRepository{
    @Transactional
    override fun findAll(): List<Actor> {
        return ActorTable
            .selectAll()
            .map { fromRow(it) }
    }

    private fun fromRow(row: ResultRow): Actor {
        return Actor(row[ActorTable.id], row[ActorTable.name])
    }
}
```

**Reply by micheljung** [reply]
Unfortunately, no. I dropped Exposed (I'm not a fan of ORM anyway) and use [Reactive SQL Clients](https://quarkus.io/guides/reactive-sql-clients) now.

**Reply by florentsorel** [reply]
@micheljung FYI, I found a solution, but I can't use `@Transactional` annotation. One day I hope I'll find a way to use `@Transactional` but it's not today :(

```kotlin
class ActorRepositoryImpl : ActorRepository {
    override fun findAll(): List<Actor> {
        return transaction {
            ActorTable
                .selectAll()
                .map { fromRow(it) }
        }
    }
}
```

**Reply by micheljung** [reply]
@florentsorel yes, that's what you're forced to do. But if you want to make 2 repository calls in the same transaction, you're out of luck. Then you need to wrap both calls in a `transaction { }`. Where would you do that? In the service? In the HTTP endpoint? Whatever you choose, this layer will be polluted with knowledge about Exposed - something that should be an implementation detail of your Repository ☹️ This I wanted to avoid.

**Reply by florentsorel** [reply]
@micheljung And the @Transacionnal annotation fix what you said? 🤔 
Maybe I'll try Ktorm, I just want a query builder not ORM and I like Exposed DSL approach but I think I need someone smarter than me 😞

**Reply by mschorsch** [reply]
> ... I just want a query builder not ORM ...

I would recommend [JOOQ](https://www.jooq.org/). It's easy to use, generates typesafe SQL and works out of the box with JTA (e.g. Transactional).

**Reply by micheljung** [reply]
@florentsorel if `@Transactional` worked with Exposed, you could use it e.g. in your HTTP endpoint and have your entire request run in a single transaction. Because `@Transactional` is Jakarta EE API, it's technology-neutral, so the underlying DB client or ORM wouldn't matter and can easily be replaced without touching the way you do transactions. With `transaction { }` from exposed, this is not possible.

**Reply by CXwudi** [reply]
Maybe this can be the workaround, https://github.com/CXwudi/realworld-compose-http4k-example-app/blob/master/conduit-backend/src/main/kotlin/mikufan/cx/conduit/backend/db/TransactionManager.kt
